app/models/pii/scrubber/scrub_model_pii.rb

```diff
@@ -0,0 +1,114 @@
+###
+# Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+require 'progress_bar'
+
+module Pii::Scrubber
+  # Scrub personally identifiable information (PII) from the given scope
+  # usage:
+  #   scrubber = ScrubModelPii.new(progress: true)
+  #   scrubber.perform(Hmis::Hud::Client.all)
+  class ScrubModelPii
+    attr_accessor :custom_scrubber
+    def initialize(progress: false, custom_scrubber: nil)
+      @progress = progress
+
+      self.custom_scrubber = lookup_custom_scrubber(custom_scrubber)
+    end
+
+    def perform(scope)
+      model = scope.klass
+      raise "#{model.name} is missing pii attribute configuration" unless model.stores_pii?
+
+      total = scope.count
+      return total if total.zero?
+
+      bar = new_progress_bar(total) if @progress
+      bar.puts model.name if @progress
+      without_optimistic_locking(model) do
+        process_model(model, scope, bar)
+      end
+      total
+    end
+
+    protected
+
+    def process_model(model, scope, progress)
+      non_nullable_cols = model.columns.reject(&:null).map { |c| c.name.to_sym }
+      scope.find_in_batches do |batch|
+        values = batch.map do |record|
+          pii_fields = Pii::Scrubber::PiiAttribute.from_record(record)
+
+          # custom scrubbing if provided (fake values)
+          custom_scrubber&.perform(pii_fields)
+
+          # handle dob and age fields
+          dob_scrubber.perform(pii_fields.reject(&:scrubbed?))
+
+          # scrub remaining sensitive fields
+          basic_scrubber.perform(pii_fields.filter { |f| f.sensitive? && !f.scrubbed? })
+
+          pii_attrs = pii_fields.
+            filter(&:scrubbed?).
+            to_h { |f| [f.name, f.scrubbed_value] }
+
+          # include non-nullable attrs to ensure upsert works in postgres
+          required_attrs = non_nullable_cols.to_h do |column|
```

Comment:
Why the `to_h` here?  should it be after the `end`?

Reply:
this is a variant of array.to_h that takes a block returning [key,value]
```
array.to_h { |item| [key, value] }
=> { key => value}
```

Which accomplishes the same thing as `array.map { |i| [k,v] }.to_h` but without an intermediate array.

Reply:
Amazing.  Thanks.